core/src/main/java/org/apache/druid/math/expr/BinaryLogicalOperatorExpr.java

```diff
@@ -385,7 +322,59 @@ protected BinaryOpExprBase copy(Expr left, Expr right)
   public ExprEval eval(ObjectBinding bindings)
   {
     ExprEval leftVal = left.eval(bindings);
-    return leftVal.asBoolean() ? right.eval(bindings) : leftVal;
+    if (!ExpressionProcessing.useStrictBooleans()) {
+      return leftVal.asBoolean() ? right.eval(bindings) : leftVal;
+    }
+
+    // if left is false, always false
+    if (leftVal.value() != null && !leftVal.asBoolean()) {
+      return ExprEval.ofLongBoolean(false);
```

Comment:
This looks like a constant. Can it be defined as such? `static final ExprEval.FALSE = ExprEval.ofLongBoolean(false)` to avoid recomputing the value in an inner loop? Here and below.

Reply:
yeah, it seems reasonable to make true and false constants, I may make this change in this PR or in a follow-up later if nothing else comes up that needs immediate change